### PR TITLE
WIP - language definition consistency rules

### PR DESCRIPTION
**Pull request by kurt-r2c**
Adding some rules to autofix Semgrep rules with consistent language definitions in the `language` rule definition key.

Some open questions:

- should we standardize on `terraform` or `hcl`? Terraform gurus may prefer `hcl`, people with less domain expertise will probably find `terraform` more accessible
- `c++` vs `cpp`? `c++` is "more correct", `cpp` is more HTTP-friendly

FYI @pieterdc1

### CODE REVIEW

**Comment by ievans** [comment]
Do you want to make the "canonical" name defined in https://github.com/returntocorp/semgrep-langs and use that to compile these rules?

**Comment by kurt-r2c** [comment]
per @ievans suggestion - using `semgrep-langs` as "source of truth" for lang configs. Once this is approved + merged, will open second PR with autofixes applied.

File: yaml/semgrep/consistency/lang-consistency-typescript.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+              languages: ...
+      - metavariable-regex:
+          metavariable: $X
+          regex: ^typescript$
+    message: >-
+      Found '$X' in language config which diverges from semgrep.dev normalization.
+    languages:
+      - yaml
+    severity: WARNING
+    fix: ts
```
Comment (minusworld) [inline yaml/semgrep/consistency/lang-consistency-typescript.yaml]: Most of the other languages are spelled out - I don't care which we choose, but is it more consistent to spell this and JavaScript out as well?

Comment (kurt-r2c) [reply yaml/semgrep/consistency/lang-consistency-typescript.yaml]: I agree with this, however semgrep.dev does not: 
https://github.com/returntocorp/semgrep-langs/blob/2cb383112a7780e5dfe4d768161c3d4ddb0bac57/lang.json#L229-L231

Comment (minusworld) [reply yaml/semgrep/consistency/lang-consistency-typescript.yaml]: Hm, gotcha 👍

**Review by minusworld — APPROVED** [review]

**Comment by kurt-r2c** [comment]
@minusworld @ievans now with more autogeneration from `semgrep-rules`

File: yaml/semgrep/consistency/lang-consistency-cpp.yaml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  metadata:
+    category: correctness
+    technology:
+    - semgrep
+  patterns:
+  - pattern: $X
+  - pattern-inside: 'languages: ...'
+  - metavariable-regex:
+      metavariable: $X
+      regex: ^(c++)$
```
Comment (ievans) [inline yaml/semgrep/consistency/lang-consistency-cpp.yaml]: does this need to be escaped?

File: yaml/semgrep/consistency/generate_rules.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                        },
+                        {"pattern-inside":"languages: ..."},
+                        {
+                            "metavariable-regex": {
+                                "metavariable":"$X",
+                                "regex":f"^({key_string})$"
+                            }
+                        }
+                    ],
+                    "message": "Found '$X' in language config which diverges from semgrep.dev normalization.",
```
Comment (ievans) [inline yaml/semgrep/consistency/generate_rules.py]: Maybe say: `"$X is a valid langauge key but for standardization please use " + lang_data['id']"` just a bit clearer to me even though we already have `fix`

File: yaml/semgrep/consistency/lang-consistency-cpp.yaml (hunk 2 above)
Comment (kurt-r2c) [reply yaml/semgrep/consistency/lang-consistency-cpp.yaml]: yep. will fix in generator script

File: yaml/semgrep/consistency/generate_rules.py (hunk 3 above)
Comment (kurt-r2c) [reply yaml/semgrep/consistency/generate_rules.py]: fair point will fix

**Comment by kurt-r2c** [comment]
@ievans pretty sure I got all your comments, please sanity check to make sure

**Review by ievans — APPROVED** [review]
Lgtm! Now let's get this working via semgrep-app on semgrep-rules repository, so we get those sweet new autofix comments!

**Pull request merged**